walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -1,23 +1,102 @@
-//
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at
-//
-//     https://www.apache.org/licenses/LICENSE-2.0
-//
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
 
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    //General Outline:
+    //  1. Go through each person's events and add it to a list that holds all 
+    //      times when people can't meet
+    //  2. Sort that list by start time
+    //  3. Merge overlapping intervals if any so that start/end times are clear
+    //  4. Start from beginning of day to end and the result will be all the times in this 
+    //      merged timeranges that are atleast the requested meeting's duration
+
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()){
+      return Arrays.asList();
+   
+    } else if (events.isEmpty()){
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    ArrayList<TimeRange> allUnavailableTimes = new ArrayList<>();
+
+    for (Event event : events){
+      if (validatePerson(event, request)){
+        allUnavailableTimes.add(event.getWhen());
+      }
+    }
+
+    Collections.sort(allUnavailableTimes, TimeRange.ORDER_BY_START);
+
+    ArrayList<TimeRange> mergedUnavailableTimes = mergeTimeRanges(allUnavailableTimes);
+    ArrayList<TimeRange> availableTimes = new ArrayList<>();
+    
+    TimeRange possibleTime = TimeRange.fromStartDuration(-1,-1); //junk value
+    int prevEndTime = TimeRange.START_OF_DAY;  //junk value
+    
+    for (TimeRange curTime: mergedUnavailableTimes){
+
+      possibleTime = TimeRange.fromStartEnd(prevEndTime, curTime.start(), false);
+
+      if (possibleTime.duration() >= request.getDuration()){
+        availableTimes.add(possibleTime);
+      }
+
+      prevEndTime = curTime.end();
+    }
+
+    //check to see if there's a possible time after the end of the last meeting
+    possibleTime = TimeRange.fromStartEnd(prevEndTime, TimeRange.END_OF_DAY, true);
+    if (possibleTime.duration() >= request.getDuration()){
+      availableTimes.add(possibleTime);
+    }
+
+    return availableTimes;
+  }
+
+  //merges overlapping intervals so that start and end times are clear and consecutive
+  public ArrayList<TimeRange> mergeTimeRanges(ArrayList<TimeRange> allUnavailableTimes){
```

Comment:
I think that allUnavailableTimes needs to be sorted for this function to work properly.  You should either include that in the function comment, or better yet, sort the array in this function instead of before its called

Reply:
Yeah sorting it inside is a good idea!